### PR TITLE
Add Form Component

### LINKED ISSUE #35: Implement Form component
_Estimated Delivery Sprint - Sprint 3_

- [x] Research
- [x] Specification
- [x] Component and tests implementation
- [x] API Documentation

### PR DESCRIPTION
**Pull request by igdmdimitrov**
Closes #35

### CODE REVIEW

File: src/components/form/form.ts [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+    formElements.forEach((element) => {
+      if (
+        (this._controlsWithChecked.includes(element.tagName.toLowerCase()) &&
+          element.checked) ||
+        (this._controlsWithValue.includes(element.tagName.toLowerCase()) &&
+          element.type !== 'checkbox' &&
+          element.type !== 'radio' &&
+          element.type !== 'submit')
+      ) {
+        formData.append(element.name, element.value);
```
Comment (DiyanDimitrov) [inline src/components/form/form.ts:99]: We have to handle `select` element with a `multiple` attribute in a special way.

File: src/components/form/form.ts [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+    return formElements;
+  }
+
+  getFormData() {
+    const formData = new FormData();
+
+    const formElements = this.getFormElements();
+    formElements.forEach((element) => {
+      if (
+        (this._controlsWithChecked.includes(element.tagName.toLowerCase()) &&
```
Comment (DiyanDimitrov) [inline src/components/form/form.ts:92]: We have to skip disabled elements.

File: src/components/form/form.ts [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+      targetElement.type.toLowerCase() === 'submit'
+    ) {
+      const formData = this.getFormData();
+
+      const isValid = this.reportValidity();
+      if (!this.novalidate && !isValid) {
+        return false;
+      }
+
+      this.emitEvent('igcSubmit', { detail: formData });
```
Comment (DiyanDimitrov) [inline src/components/form/form.ts:128]: Shouldn't we just call the submit method?

**Review by DiyanDimitrov — CHANGES_REQUESTED** [review]

File: stories/form.stories.ts [hunk 4]
```diff
@@ -0,0 +1,133 @@
+import { html } from 'lit-html';
+import '../igniteui-webcomponents.js';
+import { Story } from './story.js';
+// import { ifDefined } from 'lit-html/directives/if-defined';
```
Comment (zhosti) [inline stories/form.stories.ts:4]: Remove the commented code

**Review by zhosti — APPROVED** [review]

File: src/components/form/form.ts [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+    const formData = new FormData();
+
+    const formElements = this.getFormElements();
+    formElements.forEach((element) => {
+      const tagName = element.tagName.toLowerCase();
+      if (tagName === 'select') {
+        for (let i = 0; i < element.options.length; i++) {
+          const option = element.options[i];
+          if (option.selected) {
+            formData.append(element.name, option.text);
```
Comment (DiyanDimitrov) [inline src/components/form/form.ts:116]: We should use the `option.value`.
```suggestion
            formData.append(element.name, option.value);
```

**Review by DiyanDimitrov — CHANGES_REQUESTED** [review]

File: src/components/form/form.ts [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+    super();
+
+    this.addEventListener('click', this.handleClick);
+  }
+
+  /** Submits the form. */
+  submit(): boolean {
+    const formData = this.getFormData();
+
+    const isValid = this.reportValidity();
```
Comment (DiyanDimitrov) [inline src/components/form/form.ts:55]: `reportValidity` should not be called if `novalidate` is true.

File: src/components/form/form.ts [hunk 7]
```diff
@@ -0,0 +1,158 @@
...
+  `;
+
+  private _controlsWithChecked = [
+    'input',
+    'radio',
+    'igc-radio',
+    'igc-switch',
+    'igc-checkbox',
+  ];
+  private _controlsWithValue = ['input', 'textarea'];
```
Comment (DiyanDimitrov) [inline src/components/form/form.ts:39]: Add `igc-input` here. Also add `igc-input` and `igc-checkbox` to the story.

File: src/components/form/form.ts [hunk 8]
```diff
@@ -0,0 +1,164 @@
...
+      } else if (
+        (tagName === 'input' &&
+          (element.type === 'checkbox' || element.type == 'radio')) ||
+        (tagName !== 'input' && this._controlsWithChecked.includes(tagName))
+      ) {
+        element.checked = element.hasAttribute('checked');
+      } else if (tagName === 'igc-input') {
+        element.value = element.getAttribute('value');
+      } else if (this._controlsWithValue.includes(tagName)) {
+        element.value = element.defaultValue;
```
Comment (DiyanDimitrov) [inline src/components/form/form.ts:81]: For consistency we could use `getAttribute` for all value controls.
```suggestion
      } else if (this._controlsWithValue.includes(tagName)) {
        element.value = element.getAttribute('value');
```

**Review by DiyanDimitrov — CHANGES_REQUESTED** [review]

File: src/components/form/form.ts (hunk 8 above)
Comment (DiyanDimitrov) [reply src/components/form/form.ts]: Aaand we broke the textarea... You could revert this change :)

File: src/components/form/form.ts [hunk 9]
```diff
@@ -0,0 +1,164 @@
...
+    'input',
+    'radio',
+    'igc-radio',
+    'igc-switch',
+    'igc-checkbox',
+  ];
+  private _controlsWithValue = ['input', 'igc-input', 'textarea'];
+  private _controlsThatSubmit = ['input', 'button', 'igc-button'];
+
+  /** Specifies if the form should not to be validated on submit. */
```
Comment (gedinakova) [inline src/components/form/form.ts:42]: Let's change this to something like "Specifies if form data validation should be skipped on submit."

**Review by gedinakova — CHANGES_REQUESTED** [review]

**Review by gedinakova — APPROVED** [review]

**Review by DiyanDimitrov — APPROVED** [review]

**Review by zhosti — APPROVED** [review]

**Pull request merged**